### PR TITLE
feat: lru-k implementation

### PR DESCRIPTION
**Pull request by unw9527**
This PR implements the LRU-K algorithm described in the paper [The LRU-K page replacement algorithm for database disk buffering](https://doi.org/10.1145/170036.170081) for the cache. Unit tests are included.

ParpulseCache, which is the one exposed to Storage Manager, is another layer of abstraction atop. Thus, previously/subsequently implemented cache replacement policies will not be affected.

### CODE REVIEW

File: storage-node/src/cache/lru_k.rs [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+            }
+            let history = &node.history;
+            if let Some(kth_timestamp) = history.front() {
+                k_dist = if history.len() < self.k {
+                    std::i32::MAX
+                } else {
+                    self.curr_timestamp - kth_timestamp
+                };
+                if (k_dist > max_k_dist)
+                    || (k_dist == max_k_dist && *kth_timestamp < earliest_timestamp)
```
Comment (xx01cyx) [inline storage-node/src/cache/lru_k.rs]:
```suggestion
                if (k_dist > max_k_dist)
                    || (k_dist == max_k_dist && kth_timestamp < &earliest_timestamp)
```
A very small point: we can save one copy if we use `&earliest_timestamp` for comparison.

File: storage-node/src/cache/lru_k.rs [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+        self.curr_timestamp += 1;
+    }
+
+    fn get_value(&mut self, key: &K) -> Option<&V> {
+        if let Some(mut node) = self.cache_map.remove(key) {
+            self.record_access(&mut node);
+            self.cache_map.insert(key.clone(), node);
+            if let Some(cache_value) = self.cache_map.get(key).map(|node| &node.value) {
+                return Some(cache_value);
+            }
```
Comment (xx01cyx) [inline storage-node/src/cache/lru_k.rs]:
```suggestion
            return self.cache_map.get(key).map(|node| &node.value);
```

File: storage-node/src/cache/lru_k.rs [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+                },
+            );
+            self.curr_timestamp += 1;
+        }
+        self.size += updated_size;
+        while self.size > self.max_capacity {
+            let key_to_evict = self.evict(&key);
+            if key_to_evict.is_none() {
+                panic!("Fail to evict key {:?}", key);
+            }
```
Comment (xx01cyx) [inline storage-node/src/cache/lru_k.rs]:
```suggestion
            debug_assert!(
                key_to_evict.is_some(),
                "key {:?} should be evicted when cache size is greater than max capacity",
                key
            );
```
Maybe we can make the message clearer.

**Review by xx01cyx — COMMENTED** [review]
Shall we also test the case where there are keys with infinite kth distance in the cache? In that case, the key with the earliest overall timestamp should be evicted.

**Comment by unw9527** [comment]
> Shall we also test the case where there are keys with infinite kth distance in the cache? In that case, the key with the earliest overall timestamp should be evicted.

Added.

**Review by xx01cyx — APPROVED** [review]
LGTM

**Pull request merged**